system_setup/server/controllers/shutdown.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2020-2021 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from subiquitycore.context import with_context
+from subiquity.common.types import ShutdownMode
+from subiquity.server.controllers import ShutdownController
+
+log = logging.getLogger("system_setup.controllers.restart")
+
+
+class SetupShutdownController(ShutdownController):
```

Comment:
All the event management is inherited and in line with what subiquity is doing, we would prefer to keep the inheritance rather than reimplementing this code.